### DISCUSSION TITLE
How do I use constraints to represent basic mathematical types

### DISCUSSION THREAD

**Discussion by zms9110750** [dotnet/runtime#102614] [Q&A]
I want a method that can be applied to bytes,  sbyte, short, ushort, ... .
The INumber interface seems to only specify how to perform operations on its own type. Unable  with constant int, double type interaction.

The preset basic type is achieved through conversion.
But this interface does not mention conversion.

**Comment by elgonzo** [comment]
> But this interface does not mention conversion.

`INumber<T>` inherits several other interfaces. Check them out. For example, conversion are done through the static `Create*` methods specified by the `INumberBase<T>` interface.

> How do I use constraints to represent basic mathematical types?

This depends on what your generic method should actually do. If you only want to do arithmethic on either integer types or either floating point types, use `IBinaryInteger<T>` or `IFloatingPoint<T>` as constraint.

<br>

If however you want to do also arithmethic between integers and floating point types, `INumber<T>` is a suitable constraint. But this typically means floating point arithmethic which in turn introduces issues of precision and possible incidental value range violations when you want integer results. (In such cases, it's often a good idea to have a second generic method constrained to integer types only (IBinaryInteger\<T\>) in which you can then do accurate `Int128`-based integer arithmetic (unless you intend to also do math with BigIntegers...)
